Ensure SafeFontParser is used more widely.

robert-jenner opened the pull request:
#### 524b5e70e865c270e631735d67a175682ed6b58f
<pre>
Ensure SafeFontParser is used more widely.
<a href="https://bugs.webkit.org/show_bug.cgi?id=299511">https://bugs.webkit.org/show_bug.cgi?id=299511</a>
<a href="https://rdar.apple.com/160955353">rdar://160955353</a>

Reviewed by Vitor Roriz and Brent Fulgham.

SafeFontParser is adopted either when we enable Lockdown Mode, or when
an explicit setting is set to force it to be used. There were several
cases where we still did not use SafeFontParser for custom fonts - when
creating a font request in a Worker, and primarily when supplying
immediate font sources that did not require a network load.

This adopts SafeFontParser for these cases too and adjusts several calls
to be more explicitly clear in their function.

Testing this is quite tricky as we do not expose anything about being
handled by SafeFontParser. Instead, currently, a canary font is used in
two additional test cases that should fail to parse in SafeFontParser,
but succeeds in FontParser. This allows a proxy measure for identifying
if we&apos;ve been loaded in SafeFontParser.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::CSSFontFace):
(WebCore::CSSFontFace::opportunisticallyStartFontDataURLLoading):
(WebCore::CSSFontFace::pump):
(WebCore::CSSFontFace::font):
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::opportunisticallyStartFontDataURLLoading):
(WebCore::CSSFontFaceSource::loadCustomFont):
(WebCore::CSSFontFaceSource::load):
* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::opportunisticallyStartFontDataURLLoading):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::ensureCustomFontData):
(WebCore::CachedFont::createCustomFontData):
(WebCore::CachedFont::createCustomFontDataWithPolicy):
(WebCore::CachedFont::createCustomFontDataSystemParser):
(WebCore::CachedFont::createCustomFontDataSafeFontParser):
(WebCore::CachedFont::createCustomFontDataExperimentalParser): Deleted.
* Source/WebCore/loader/cache/CachedFont.h:
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::create):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::loadCustomFont):
(WebCore::WorkerFontLoadRequest::ensureCustomFontData):
* Source/WebCore/workers/WorkerFontLoadRequest.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.mm:
(TestWebKitAPI::EncodeResourceAsArrayString):
(TestWebKitAPI::TEST(LockdownMode, NotAllowedFontLoadingAPI)):
(TestWebKitAPI::TEST(LockdownMode, AllowedFontLoadingAPI)):
(TestWebKitAPI::TEST(LockdownMode, NotSupportedFontLoadingAPI)):
(TestWebKitAPI::TEST(LockdownMode, DISABLED_ImmediateParsedViaSafeFontParser)):
(TestWebKitAPI::TEST(LockdownMode, DISABLED_CanaryFontSucceedsInFontParser)):
(TestWebKitAPI::TEST(LockdownMode, DISABLED_WorkerFontParsedViaSafeFontParser)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeFontParser-Invalid.ttf: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeFontParser-Worker.html: Added.

Originally-landed-as: 297297.479@safari-7622-branch (0d4976340f1d). <a href="https://rdar.apple.com/164212005">rdar://164212005</a>
Canonical link: <a href="https://commits.webkit.org/304167@main">https://commits.webkit.org/304167@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6ac94f307310de24a0146fc40c8bd7395c5b9fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134661 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7118 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45900 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142185 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86602 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e05d470c-4140-40b5-b49a-a19466ba67ea) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136531 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7718 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6971 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102920 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70203 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5b945b24-c037-49b6-bf5b-96f3caea34d2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5413 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120699 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83725 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ace70444-e2ad-4807-a8c3-056dbb937918) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5269 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2885 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2776 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38839 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144878 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6792 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39420 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111315 "Found 1 new test failure: fast/images/async-image-multiple-clients-repaint.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6866 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5695 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111610 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28335 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5108 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116972 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60676 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6843 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35164 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6648 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70424 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6884 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6757 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt reviewed (commented):
This already has reviewed by Vitor and Brent. Are we re-landing this patch?

Commit:
Ensure SafeFontParser is used more widely.
https://bugs.webkit.org/show_bug.cgi?id=299511
rdar://160955353

Reviewed by Vitor Roriz and Brent Fulgham.

SafeFontParser is adopted either when we enable Lockdown Mode, or when
an explicit setting is set to force it to be used. There were several
cases where we still did not use SafeFontParser for custom fonts - when
creating a font request in a Worker, and primarily when supplying
immediate font sources that did not require a network load.

This adopts SafeFontParser for these cases too and adjusts several calls
to be more explicitly clear in their function.

Testing this is quite tricky as we do not expose anything about being
handled by SafeFontParser. Instead, currently, a canary font is used in
two additional test cases that should fail to parse in SafeFontParser,
but succeeds in FontParser. This allows a proxy measure for identifying
if we've been loaded in SafeFontParser.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::CSSFontFace):
(WebCore::CSSFontFace::opportunisticallyStartFontDataURLLoading):
(WebCore::CSSFontFace::pump):
(WebCore::CSSFontFace::font):
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::opportunisticallyStartFontDataURLLoading):
(WebCore::CSSFontFaceSource::loadCustomFont):
(WebCore::CSSFontFaceSource::load):
* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::opportunisticallyStartFontDataURLLoading):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::ensureCustomFontData):
(WebCore::CachedFont::createCustomFontData):
(WebCore::CachedFont::createCustomFontDataWithPolicy):
(WebCore::CachedFont::createCustomFontDataSystemParser):
(WebCore::CachedFont::createCustomFontDataSafeFontParser):
(WebCore::CachedFont::createCustomFontDataExperimentalParser): Deleted.
* Source/WebCore/loader/cache/CachedFont.h:
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::create):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::loadCustomFont):
(WebCore::WorkerFontLoadRequest::ensureCustomFontData):
* Source/WebCore/workers/WorkerFontLoadRequest.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.mm:
(TestWebKitAPI::EncodeResourceAsArrayString):
(TestWebKitAPI::TEST(LockdownMode, NotAllowedFontLoadingAPI)):
(TestWebKitAPI::TEST(LockdownMode, AllowedFontLoadingAPI)):
(TestWebKitAPI::TEST(LockdownMode, NotSupportedFontLoadingAPI)):
(TestWebKitAPI::TEST(LockdownMode, DISABLED_ImmediateParsedViaSafeFontParser)):
(TestWebKitAPI::TEST(LockdownMode, DISABLED_CanaryFontSucceedsInFontParser)):
(TestWebKitAPI::TEST(LockdownMode, DISABLED_WorkerFontParsedViaSafeFontParser)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeFontParser-Invalid.ttf: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SafeFontParser-Worker.html: Added.

Originally-landed-as: 297297.479@safari-7622-branch (0d4976340f1d). rdar://164212005
Canonical link: https://commits.webkit.org/304167@main
